### PR TITLE
Improvements based on trying on Jenkins core

### PR DESCRIPTION
**Pull request by timja**

### CODE REVIEW

File: fetch_issues.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env python3
```
Comment (timja) [inline fetch_issues.py:1]: a script that can fetch all issues from a JQL query.
Exporting 11,000 issues for jenkins core via web-ui wasn't very fun especially given how slow the export was so I scripted it, works nicely

File: fetch_labels.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env python3
```
Comment (timja) [inline fetch_labels.py:1]: useful for creating an allowlist of labels. people put a lot of junk in Jira labels as there's no permissions required

File: importer.py [hunk 3]
```diff
@@ -131,6 +151,7 @@ def import_issues(self, start_from_count):
 
         for issue in self.project.get_issues():
             if start_from_count > count:
+                count += 1
```
Comment (timja) [inline importer.py:154]: `start_from_count` seemed broken, it did nothing before this fix.

File: importer.py [hunk 4]
```diff
@@ -242,7 +262,7 @@ def convert_relationships_to_comments(self, issue):
 
         for duplicate_item in duplicates:
             issue['comments'].append(
-                {"body": "<i>[Duplicates: " + self._replace_jira_with_github_id(duplicate_item)})
+                {"body": '<i>[Duplicates: <a href="https://github.com/jenkins-infra/helpdesk/issues?q=' + self._replace_jira_with_github_id(duplicate_item) + '">' + self._replace_jira_with_github_id(duplicate_item) + '</a>]</i>'})
```
Comment (timja) [inline importer.py:265]: this outputted `[Duplicates $jira_id`
before, seemed half finished

File: main.py [hunk 5]
```diff
@@ -1,30 +1,22 @@
...
-
-jira_proj = input('JIRA project name: ') or 'INFRA'
-jira_done_id = input('JIRA Done statusCategory ID [default "3"]: ') or '3'
-jira_base_url = input('JIRA base url [default "https://issues.jenkins.io"]: ') or 'https://issues.jenkins.io'
-ac = input('GitHub account name (user/org): ') or 'jenkins-infra'
-repo = input('GitHub repository name: ') or 'helpdesk'
-pat = input('Github Personal Access Token: ') # or '<your-github-pat>'
+from utils import read_xml_files
+
+file_names = os.getenv('JIRA_MIGRATION_FILE_PATHS') or input(
```
Comment (timja) [inline main.py:10]: very annoying running this over and over again with prompts, I changed it to default to env variables

File: project.py [hunk 6]
```diff
@@ -93,7 +98,8 @@ def _append_item_to_project(self, item):
 
         body = self._htmlentitydecode(item.description.text)
         # metadata: original author & link
-        body = body + '\n\n---\n<details><summary><i>Originally reported by <a title="' + item.reporter + '" href="' + self.jiraBaseUrl + '/secure/ViewProfile.jspa?name=' + item.reporter.get('username') + '">' + item.reporter.get('username') + '</a>, imported from: <a href="' + self.jiraBaseUrl + '/browse/' + item.key.text + '" target="_blank">' + item.title.text[item.title.text.index("]") + 2:len(item.title.text)] + '</a></i></summary>'
+
+        body = body + '\n\n---\n<details><summary><i>Originally reported by <a title="' + str(item.reporter) + '" href="' + self.jiraBaseUrl + '/secure/ViewProfile.jspa?name=' + item.reporter.get('username') + '">' + item.reporter.get('username') + '</a>, imported from: <a href="' + self.jiraBaseUrl + '/browse/' + item.key.text + '" target="_blank">' + item.title.text[item.title.text.index("]") + 2:len(item.title.text)] + '</a></i></summary>'
```
Comment (timja) [inline project.py:102]: Someone with a username with only numbers caused this code to fail, casting it to a string fixed it

File: project.py [hunk 7]
```diff
@@ -120,12 +126,11 @@ def _append_item_to_project(self, item):
...
         for component in item.component:
-            labels.append('jira-component:' + component.text.lower())
-            labels.append(component.text.lower())
+            if os.getenv('JIRA_MIGRATION_INCLUDE_COMPONENT_IN_LABELS', 'true') == 'true':
+                labels.append('jira-component:' + component.text.lower())
+                labels.append(component.text.lower())
 
-        #labels.append('type:' + component.text)
-        for ltype in self._project['Types'].keys():
-            labels.append('jira-type:' + component.text)
```
Comment (timja) [inline project.py:128]: there was a typo here that it used the component rather than type

File: project.py [hunk 8]
```diff
@@ -120,12 +126,11 @@ def _append_item_to_project(self, item):
...
+        labels.append(self._jira_type_mapping(item.type.text.lower()))
```
Comment (timja) [inline project.py:133]: instead of prefixing with `jira-type` map expected types to an expected value

File: project.py [hunk 9]
```diff
@@ -155,6 +160,22 @@ def _append_item_to_project(self, item):
         if not self._project['Issues'][-1]['closed_at']:
             del self._project['Issues'][-1]['closed_at']
 
+    def _jira_type_mapping(self, issue_type):
```
Comment (timja) [inline project.py:163]: could move this to a config file if required

**Review by timja — COMMENTED** [review]
Happy to clean this up more and add some docs if you think it's worth it.

I've ran this on ~3500 issues so far and found a few issues with the existing code that I've cleaned up

**Comment by timja** [comment]
cc @lemeurherve not sure if you've seen this.

I added epics support in https://github.com/lemeurherve/jira-issues-importer/pull/3/commits/a3f4237e90f90216c008153cfbf8ce690bff9638 :)

File: post_process_issues.sh [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+ALL_ISSUES=$(gh issue list -R ${OWNER}/${REPO} --limit=20000 --state=all --json number,labels)
+ALL_ISSUE_NUMBERS=$(echo "${ALL_ISSUES}"| jq '.[].number' | sort -g | uniq)
+
+while IFS= read -r ISSUE_CHECKING; do
+  if (( ISSUE_CHECKING < START_FROM )); then
+    continue
+  fi
+  echo "Checking $ISSUE_CHECKING"
+  COMMENT=$(gh issue view -R timja/jenkins-gh-issues-poc-06-18 "${ISSUE_CHECKING}" --comments --json 'comments' --jq '.comments[].body | select(contains("[Epic:"))')
```
Comment (lemeurherve) [inline post_process_issues.sh:20]: Nice!

File: post_process_issues.sh [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+Epic children:
+
+- #${ISSUE_CHECKING}"
+      fi
+      gh issue edit -R ${OWNER}/${REPO} "${EPIC_ISSUE_NUMBER}" --body "${BODY}"
+    fi
+
+  fi
+
+done <<< "${ALL_ISSUE_NUMBERS}"
```
Comment (lemeurherve) [inline post_process_issues.sh:52]:
```suggestion
done <<< "${ALL_ISSUE_NUMBERS}"

```

File: importer.py [hunk 12]
```diff
@@ -239,32 +246,44 @@ def convert_relationships_to_comments(self, issue):
...
         blocks = issue['blocks']
+        try:
+            epic_link = issue['epic-link']
+        except (AttributeError, KeyError):
+            epic_link = None
 
         for duplicate_item in duplicates:
             issue['comments'].append(
-                {"body": "<i>[Duplicates: " + self._replace_jira_with_github_id(duplicate_item)})
+                {"body": f'<i>[Duplicates: <a href="https://github.com/{self.options.account}/{self.options.repo}/issues?q=' + self._replace_jira_with_github_id(duplicate_item) + '">' + self._replace_jira_with_github_id(duplicate_item) + '</a>]</i>'})
```
Comment (lemeurherve) [inline importer.py:256]:
```suggestion
                {"body": f'<i>[Duplicates: <a href="https://github.com/{self.options.account}/{self.options.repo}/issues?q=%22' + self._replace_jira_with_github_id(duplicate_item) + '%22">' + self._replace_jira_with_github_id(duplicate_item) + '</a>]</i>'})

```

We should add double quotes for a strict GitHub search. I missed that for INFRA issues.

Ex:
- [no quote](https://github.com/jenkins-infra/helpdesk/issues?q=is%3Aissue+INFRA-2000) (wrongly include INFRA-325 which contains the text `2000`): `https://github.com/jenkins-infra/helpdesk/issues?q=is%3Aissue+INFRA-2000`
- [simple quotes](https://github.com/jenkins-infra/helpdesk/issues?q=is%3Aissue+%27INFRA-2000%27) (same): `https://github.com/jenkins-infra/helpdesk/issues?q=is%3Aissue+%27INFRA-2000%27`
- [double quotes](https://github.com/jenkins-infra/helpdesk/issues?q=is%3Aissue+%22INFRA-2000%22): `https://github.com/jenkins-infra/helpdesk/issues?q=is%3Aissue+%22INFRA-2000%22`

File: importer.py [hunk 13]
```diff
@@ -239,32 +246,44 @@ def convert_relationships_to_comments(self, issue):
...
 
         for is_duplicated_by_item in is_duplicated_by:
             issue['comments'].append(
-                {"body": '<i>[Originally duplicated by: <a href="https://github.com/jenkins-infra/helpdesk/issues?q=' + self._replace_jira_with_github_id(is_duplicated_by_item) + '">' + self._replace_jira_with_github_id(is_duplicated_by_item) + '</a>]</i>'})
+                {"body": f'<i>[Originally duplicated by: <a href="https://github.com/{self.options.account}/{self.options.repo}/issues?q=' + self._replace_jira_with_github_id(is_duplicated_by_item) + '">' + self._replace_jira_with_github_id(is_duplicated_by_item) + '</a>]</i>'})
```
Comment (lemeurherve) [inline importer.py:260]:
```suggestion
                {"body": f'<i>[Originally duplicated by: <a href="https://github.com/{self.options.account}/{self.options.repo}/issues?q=%22' + self._replace_jira_with_github_id(is_duplicated_by_item) + '%22">' + self._replace_jira_with_github_id(is_duplicated_by_item) + '</a>]</i>'})

```

File: importer.py [hunk 14]
```diff
@@ -239,32 +246,44 @@ def convert_relationships_to_comments(self, issue):
...
 
         for relates_to_item in relates_to:
             issue['comments'].append(
-                {"body": '<i>[Originally related to: <a href="https://github.com/jenkins-infra/helpdesk/issues?q=' + self._replace_jira_with_github_id(relates_to_item) + '">' + self._replace_jira_with_github_id(relates_to_item) + '</a>]</i>'})
+                {"body": f'<i>[Originally related to: <a href="https://github.com/{self.options.account}/{self.options.repo}/issues?q=' + self._replace_jira_with_github_id(relates_to_item) + '">' + self._replace_jira_with_github_id(relates_to_item) + '</a>]</i>'})
```
Comment (lemeurherve) [inline importer.py:264]:
```suggestion
                {"body": f'<i>[Originally related to: <a href="https://github.com/{self.options.account}/{self.options.repo}/issues?q=%22' + self._replace_jira_with_github_id(relates_to_item) + '%22">' + self._replace_jira_with_github_id(relates_to_item) + '</a>]</i>'})

```

File: importer.py [hunk 15]
```diff
@@ -239,32 +246,44 @@ def convert_relationships_to_comments(self, issue):
...
 
         for depends_on_item in depends_on:
             issue['comments'].append(
-                {"body": '<i>[Originally depends on: <a href="https://github.com/jenkins-infra/helpdesk/issues?q=' + self._replace_jira_with_github_id(depends_on_item) + '">' + self._replace_jira_with_github_id(depends_on_item) + '</a>]</i>'})
+                {"body": f'<i>[Originally depends on: <a href="https://github.com/{self.options.account}/{self.options.repo}/issues?q=' + self._replace_jira_with_github_id(depends_on_item) + '">' + self._replace_jira_with_github_id(depends_on_item) + '</a>]</i>'})
```
Comment (lemeurherve) [inline importer.py:268]:
```suggestion
                {"body": f'<i>[Originally depends on: <a href="https://github.com/{self.options.account}/{self.options.repo}/issues?q=%22' + self._replace_jira_with_github_id(depends_on_item) + '%22">' + self._replace_jira_with_github_id(depends_on_item) + '</a>]</i>'})

```

File: importer.py [hunk 16]
```diff
@@ -239,32 +246,44 @@ def convert_relationships_to_comments(self, issue):
...
 
         for blocks_item in blocks:
             issue['comments'].append(
-                {"body": '<i>[Originally blocks: <a href="https://github.com/jenkins-infra/helpdesk/issues?q=' + self._replace_jira_with_github_id(blocks_item) + '">' + self._replace_jira_with_github_id(blocks_item) + '</a>]</i>'})
+                {"body": f'<i>[Originally blocks: <a href="https://github.com/{self.options.account}/{self.options.repo}/issues?q=' + self._replace_jira_with_github_id(blocks_item) + '">' + self._replace_jira_with_github_id(blocks_item) + '</a>]</i>'})
```
Comment (lemeurherve) [inline importer.py:272]:
```suggestion
                {"body": f'<i>[Originally blocks: <a href="https://github.com/{self.options.account}/{self.options.repo}/issues?q=%22' + self._replace_jira_with_github_id(blocks_item) + '%22">' + self._replace_jira_with_github_id(blocks_item) + '</a>]</i>'})

```

File: importer.py [hunk 17]
```diff
@@ -239,32 +246,44 @@ def convert_relationships_to_comments(self, issue):
...
+
+        if epic_link:
+            issue['comments'].append(
+                {"body": f'<i>[Epic: <a href="https://github.com/{self.options.account}/{self.options.repo}/issues?q=' + self._replace_jira_with_github_id(epic_link) + ' label:epic">' + self._replace_jira_with_github_id(epic_link) + '</a>]</i>'})
```
Comment (lemeurherve) [inline importer.py:276]:
```suggestion
                {"body": f'<i>[Epic: <a href="https://github.com/{self.options.account}/{self.options.repo}/issues?q=%22' + self._replace_jira_with_github_id(epic_link) + '%22 label:epic">' + self._replace_jira_with_github_id(epic_link) + '</a>]</i>'})

```

File: post_process_issues.sh (hunk 10 above)
Comment (lemeurherve) [inline post_process_issues.sh:20]:
```suggestion
  COMMENT=$(gh issue view -R ${OWNER}/${REPO} "${ISSUE_CHECKING}" --comments --json 'comments' --jq '.comments[].body | select(contains("[Epic:"))')
```

File: project.py (hunk 6 above)
Comment (lemeurherve) [reply project.py:107]: Good catch!
The same should be done for item.assignee l111, WDYT?

File: project.py [hunk 18]
```diff
@@ -1,9 +1,11 @@
+import os
```
Comment (lemeurherve) [inline project.py:1]: l128 we should set the imported date to the current one instead of the hardcoded one.

File: project.py (hunk 7 above)
Comment (lemeurherve) [reply project.py:128]: That explains why there were no diff between "jira-component:*" and "jira:type:*" labels on imported INFRA issues... 😅

File: project.py [hunk 19]
```diff
@@ -259,6 +284,11 @@ def _add_relationships(self, item):
         except KeyError:
             print('2. KeyError at ' + item.key.text)
 
+        for customfield in item.customfields.findall('customfield'):
+            if customfield.get('key') == 'com.pyxis.greenhopper.jira:gh-epic-link':
```
Comment (lemeurherve) [inline project.py:288]: I'm curious, do you have example of such Jira issue(s)? What are they for?

File: reset-migration.sh [hunk 20]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+REPO=$1
+
+# caution make sure anything you want to keep is managed in code
```
Comment (lemeurherve) [inline reset-migration.sh:5]: Maybe add a prompt on this script? ^^

**Review by lemeurherve — APPROVED** [review]
I haven't tested but it looks good, great job @timja !

File: reset-migration.sh (hunk 20 above)
Comment (timja) [reply reset-migration.sh:5]: tbh it's only really useful if you have very few issues, could just delete it. you get rate limited fairly quickly and it's slow.

when testing it's better to just delete and re-create the repo

File: project.py (hunk 19 above)
Comment (timja) [reply project.py:288]: this is the type of custom field epic links are contained in, all of these issues: https://github.com/timja/jenkins-gh-issues-poc-06-18/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3Aepic

File: project.py (hunk 6 above)
Comment (timja) [reply project.py:107]: only caught it from hitting it 👅 , yes you're right, pushed a fix for that

File: project.py (hunk 18 above)
Comment (timja) [reply project.py:1]: can you explain this a bit more?

Comment (lemeurherve) [reply project.py:1]: The "imported" date is added in the `details` of the issue, but I initially hardcoded it with `2022/01/10`
```
body = body + '\n<li><b>imported</b>: 2022/01/10'
```

Comment (lemeurherve) [reply project.py:1]: The "imported" date is added in the `details` of the issue, but I initially hardcoded it with `2022/01/10`
```
body = body + '\n<li><b>imported</b>: 2022/01/10'
```

File: importer.py (hunk 12 above)
Comment (timja) [reply importer.py:259]: doesn't seem to work, the last one still includes 2

what we want is this:
`in:title INFRA-2000`

https://github.com/jenkins-infra/helpdesk/issues?q=is%3Aissue+in%3Atitle+INFRA-2000

Comment (lemeurherve) [reply importer.py:259]: I like the fact issue mentioning the searched one are included by default, as the searched one is still easily identifiable in the returned list, and seing the other can help (I know it does help me on helpdesk and "old" issues)

Comment (lemeurherve) [reply importer.py:259]: We could also add a second link "(in title only)", WDYT?

Comment (timja) [reply importer.py:259]: It's not catching mentioned ones in your example unless i'm missing something it's just matching `INFRA(NUMBER)-2000`

I think people can easily modify the query to find others but the link as-is is for finding the matching issue?

Comment (lemeurherve) [reply importer.py:259]: This isn't a good example as the migration issue is catched in a lot of searches, I'll find better one(s) tomorrow.

File: project.py (hunk 18 above)
Comment (timja) [reply project.py:1]: Lol, fixed (and tested)

**Comment by timja** [comment]
I think good enough for now?

File: labelcolourselector.py [hunk 21]
```diff
@@ -3,8 +3,11 @@ def __init__(self, project):
         self._project = project
 
     def get_colour(self, label):
-        if (label in ('jira-type:task', 'jira-type:story', 'Task', 'Story', 'story')): return '7bc043'
-        elif (label == 'Bug'): return 'ee4035'
+        if label in ('rfe', 'story'):
+            return '7bc043'
+        elif label == 'bug':
+            return 'ee4035'
```
Comment (lemeurherve) [inline labelcolourselector.py:9]:
```suggestion
        elif label == 'bug':
            return 'ee4035'
        elif label == 'epic':
            return 'ddf4dd'
```
Last suggestion, WDYT of a different light color image for "epic" label?

Comment (lemeurherve) [reply labelcolourselector.py:9]: Ready to merge otherwise 🙂

Comment (timja) [reply labelcolourselector.py:9]: don't care 😄, planned to tweak colours after migration, happy to commit though

**Pull request merged**